subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareResourceHasher.java

```diff
@@ -0,0 +1,199 @@
+        this.delegate = delegate;
+        this.lineEndingSensitivity = lineEndingSensitivity;
+    }
+
+    @Override
+    public void appendConfigurationToHasher(Hasher hasher) {
+        delegate.appendConfigurationToHasher(hasher);
+        hasher.putString(getClass().getName());
+        hasher.putString(lineEndingSensitivity.name());
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(RegularFileSnapshotContext snapshotContext) throws IOException {
+        return lineEndingSensitivity.isCandidate(snapshotContext.getSnapshot()) ?
+            hashContent(snapshotContext) :
+            delegate.hash(snapshotContext);
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(ZipEntryContext zipEntryContext) throws IOException {
+        return lineEndingSensitivity.isCandidate(zipEntryContext) ?
+            hashContent(zipEntryContext) :
+            delegate.hash(zipEntryContext);
+    }
+
+    @Nullable
+    private HashCode hashContent(RegularFileSnapshotContext snapshotContext) throws IOException {
+        try {
+            return hashContent(new File(snapshotContext.getSnapshot().getAbsolutePath()));
+        } catch (BinaryContentDetectedException e) {
+            return delegate.hash(snapshotContext);
+        }
+    }
+
+    @Nullable
+    private HashCode hashContent(ZipEntryContext zipEntryContext) throws IOException {
+        return zipEntryContext.getEntry().withInputStream(inputStream -> {
+            try {
+                return hashContent(inputStream);
+            } catch (BinaryContentDetectedException e) {
+                return delegate.hash(zipEntryContext);
+            }
+        });
+    }
+
+    private HashCode hashContent(InputStream inputStream) throws BinaryContentDetectedException, IOException {
+        return new LineEndingAwareInputStreamHasher().hash(inputStream);
+    }
+
+    private HashCode hashContent(File file) throws BinaryContentDetectedException, IOException {
+        try (FileInputStream inputStream = new FileInputStream(file)) {
+            return hashContent(inputStream);
+        }
+    }
+
+    private static class BinaryContentDetectedException extends Exception {
+    }
+
+    @VisibleForTesting
+    static class LineEndingAwareInputStreamHasher {
+        private static final HashCode SIGNATURE = Hashing.signature(LineEndingAwareInputStreamHasher.class);
+        int peekAhead = -1;
+
+        public HashCode hash(InputStream inputStream) throws IOException, BinaryContentDetectedException {
+            byte[] buffer = new byte[8192];
+            PrimitiveHasher hasher = Hashing.newPrimitiveHasher();
+
+            hasher.putHash(SIGNATURE);
+            while (true) {
+                int nread = read(inputStream, buffer);
+                if (nread < 0) {
+                    break;
+                }
+
+                if (checkForControlCharacters(buffer, nread)) {
+                    throw new BinaryContentDetectedException();
+                }
```

Comment:
I don't think we should use an exception here. We could simply return an `Optional.empty()` instead to signal that we did not manage to hash with line numbers ignored.

Reply:
Yeah, I wasn't very comfortable with this either.  Your idea of using `Optional` is a good one.  Changed.